// @ngrx/router/es6/router.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { ReplaySubject } from 'rxjs/ReplaySubject';
import { LocationStrategy, PlatformLocation } from '@angular/common';
import { BrowserPlatformLocation } from '@angular/platform-browser';
import { Injectable, provide } from '@angular/core';
import { stringify as stringifyQueryParams } from 'query-string';
export let Router = class Router extends ReplaySubject {
    constructor(platformStrategy) {
        super(1);
        this.platformStrategy = platformStrategy;
        platformStrategy.onPopState(event => this._update('pop'));
        const browserBaseHref = this.platformStrategy.getBaseHref();
        this._baseHref = stripTrailingSlash(stripIndexHtml(browserBaseHref));
        this._update('push');
    }
    _update(type) {
        this.next({ path: this.path(), type });
    }
    /**
     * Returns the normalized URL path.
     */
    path() {
        return this.normalize(this.platformStrategy.path());
    }
    /**
     * Given a string representing a URL, returns the normalized URL path without leading or
     * trailing slashes
     */
    normalize(url) {
        return stripTrailingSlash(_stripBaseHref(this._baseHref, stripIndexHtml(url)));
    }
    /**
     * Given a string representing a URL, returns the platform-specific external URL path.
     * If the given URL doesn't begin with a leading slash (`'/'`), this method adds one
     * before normalizing. This method will also add a hash if `HashLocationStrategy` is
     * used, or the `APP_BASE_HREF` if the `PathLocationStrategy` is in use.
     */
    prepareExternalUrl(url, query = '') {
        if (url.length > 0 && !url.startsWith('/')) {
            url = '/' + url;
        }
        return this.platformStrategy.prepareExternalUrl(url + normalizeQueryParams(normalizeQuery(query)));
    }
    /**
     * Changes the browsers URL to the normalized version of the given URL, and pushes a
     * new item onto the platform's history.
     */
    go(path, query = '') {
        this.platformStrategy.pushState(null, '', path, normalizeQuery(query));
        this._update('push');
    }
    /**
     * Changes the browsers URL to the normalized version of the given URL, and replaces
     * the top item on the platform's history stack.
     */
    replace(path, query = '') {
        this.platformStrategy.replaceState(null, '', path, normalizeQuery(query));
        this._update('replace');
    }
    /**
     * Changes the browsers query parameters. Replaces the top item on the platform's
     * history stack
     */
    search(query = '') {
        const [pathname] = this.path().split('?');
        this.replace(pathname, query);
    }
    /**
     * Navigates forward in the platform's history.
     */
    forward() {
        this.platformStrategy.forward();
    }
    /**
     * Navigates back in the platform's history.
     */
    back() {
        this.platformStrategy.back();
    }
};
Router = __decorate([
    Injectable(), 
    __metadata('design:paramtypes', [LocationStrategy])
], Router);
function _stripBaseHref(baseHref, url) {
    if (baseHref.length > 0 && url.startsWith(baseHref)) {
        return url.substring(baseHref.length);
    }
    return url;
}
function stripIndexHtml(url) {
    if (/\/index.html$/g.test(url)) {
        // '/index.html'.length == 11
        return url.substring(0, url.length - 11);
    }
    return url;
}
function stripTrailingSlash(url) {
    if (/\/$/g.test(url)) {
        url = url.substring(0, url.length - 1);
    }
    return url;
}
function normalizeQuery(query) {
    return typeof query === 'string' ? query : stringifyQueryParams(query);
}
function normalizeQueryParams(params) {
    return (params.length > 0 && params.substring(0, 1) !== '?') ? ('?' + params) : params;
}
export const ROUTER_PROVIDERS = [
    provide(Router, { useClass: Router }),
    provide(PlatformLocation, { useClass: BrowserPlatformLocation })
];
